include/openthread/thread.h

```diff
@@ -264,6 +264,22 @@ otError otThreadDiscover(otInstance *             aInstance,
  */
 bool otThreadIsDiscoverInProgress(otInstance *aInstance);
 
+/**
+ * Set the Thread Joiner Advertisement when discovering Thread network.
+ *
+ * @param[in]  aInstance A pointer to an OpenThread instance.
+ * @param[in]  aOui      The Vendor OUI value which will be included in the Joiner Advertisement.
+ * @param[in]  aAdvData  The AdvData which will be included in the Joiner Advertisement.
+ *
+ * @retval OT_ERROR_NONE         Successfully set Joiner Advertisement.
+ * @retval OT_ERROR_INVALID_ARGS Invalid AdvData.
+ *
+ */
+otError otThreadSetJoinerAdvertisement(otInstance *   aInstance,
```

Comment:
Can we consider moving/defining the Joiner Adv related APIs in `joiner.h` (since they seem to be applicable to a joiner).

Reply:
The Joiner Advertisement is not related to `joiner.h` much actually, though they seem to be related. Technically we can move the APIs into `joiner.h`, but I was hesitating to move these APIs to `joiner.h` for a couple reasons:

- moving user facing APIs to `joiner.h` also implies that the internal APIs should better be moved from `class Mle::DiscoverScanner `to `class MeshCoP::Joiner`
- the `AdvData` however needs sent out within `DiscoverScanner` eventually where the discovery request message will be sent out

As results of above two facts:
- the parameters (such as `AdvData` and `Oui`)need passed from `Joiner` to `DiscoverScanner`, so we'll have to either change the existing `DiscoverScanner` APIs or allowing `DiscoverScanner` to access all of `Joiner`'s the private members
- the DiscoverScanner code will have to be fragmented because `OPENTHREAD_CONFIG_JOINER_ENABLE` flag will be introduced to make sure the new code only included when `MeshCoP::Joiner` code is included

And more importantly:
- doing so will disable the Joiner Advertisement function when `MeshCoP::Joiner` is not enabled, which is not always desired. E.g., the application protocol may want to use Joiner Advertisement for rendezvous purpose, but do not want to use the Thread Commissioning for provisioning.

Thoughts?

Reply:
SGTM.